// cyclomatic-complexity@1.2.5/bin/src/complexity.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const complexity_1 = require("./complexity");
describe("complexity", () => {
    test("should return 1 for a simple function", () => {
        const code = `function a() {
      return 1;
    }`;
        const result = (0, complexity_1.calculateComplexity)(code);
        expect(result).toEqual([
            { name: "global", complexity: 0, line: 0 },
            { name: "a", complexity: 1, line: 1 },
        ]);
    });
    test("should return 2 for a function with a function call", () => {
        const code = `function a() {
      b();
    }`;
        const result = (0, complexity_1.calculateComplexity)(code);
        expect(result).toEqual([
            { name: "global", complexity: 0, line: 0 },
            { name: "a", complexity: 2, line: 1 },
        ]);
    });
    test("should return 3 for a function with a function call", () => {
        const code = `function a() {
      b();
      c();
    }`;
        const result = (0, complexity_1.calculateComplexity)(code);
        expect(result).toEqual([
            { name: "global", complexity: 0, line: 0 },
            { name: "a", complexity: 3, line: 1 },
        ]);
    });
    test("should return 2 for a function with a function call", () => {
        const code = `function a() {
      if (true) {
        return 1;
      }
      return 2;
    }`;
        const result = (0, complexity_1.calculateComplexity)(code);
        expect(result).toEqual([
            { name: "global", complexity: 0, line: 0 },
            { name: "a", complexity: 2, line: 1 },
        ]);
    });
    test("should return 2 for an arrow function with a function call", () => {
        const code = `const arrow = () => {
      if (true) {
        return 1;
      }
      return 2;
    }`;
        const result = (0, complexity_1.calculateComplexity)(code);
        expect(result).toEqual([
            { name: "global", complexity: 0, line: 0 },
            { name: "anonymous: arrow", complexity: 2, line: 1 }
        ]);
    });
    test("should return 0 for an empty code block", () => {
        const code = ``;
        const result = (0, complexity_1.calculateComplexity)(code);
        expect(result).toEqual([{ name: "global", complexity: 0, line: 0 }]);
    });
    test("should return 3 for a function with recursive function call", () => {
        const code = `
      "use strict";
      class C {
          constructor() {
              this.x = a || b || c;
          }
      }
      class D {
      }
      (() => {
          if (foo) {
              bar = baz || qux;
          }
    })();`;
        const result = (0, complexity_1.calculateComplexity)(code);
        expect(result).toEqual([
            { name: "global", complexity: 0, line: 0 },
            { name: "anonymous: constructor", complexity: 3, line: 4 },
            { name: "anonymous: D", complexity: 3, line: 10 },
        ]);
    });
});
